runtime/src/main/cpp/Runtime.cpp

```diff
@@ -108,11 +96,16 @@ RuntimeState* initRuntime() {
     InitOrDeinitGlobalVariables(INIT_GLOBALS, result->memoryState);
   }
   InitOrDeinitGlobalVariables(INIT_THREAD_LOCAL_GLOBALS, result->memoryState);
+  RuntimeAssert(result->status == RuntimeStatus::kUninitialized, "Runtime must still be in the uninitialized state");
+  result->status = RuntimeStatus::kRunning;
   return result;
 }
 
 void deinitRuntime(RuntimeState* state) {
-  ResumeMemory(state->memoryState);
+  RuntimeAssert(state->status == RuntimeStatus::kRunning, "Runtime must be in the running state");
+  state->status = RuntimeStatus::kDestroying;
+  // This may be called after TLS is zeroed out, so ::memoryState in Memory cannot be trusted.
+  RestoreMemory(state->memoryState);
```

Comment:
Thing to consider when reworking this: there is a chance that this and further TLS accesses cause memory leak, at least on some platforms.
If TLS is zeroed out, then it is also deallocated. So an access to TLS allocates it again, and I'm not sure thread deinitialization machinery would take care of it.

Reply:
I agree. pthread's TLS behaviour (in zeroing out TLS before running `pthread_key_t` destructors) turns out to be rather unhelpful for us.